external/llvm-project/mlir/lib/Conversion/XModelToGPU/XModelToGPU.cpp

```diff
@@ -91,45 +91,74 @@ struct LaunchRewritePattern : public OpRewritePattern<async::LaunchOp> {
     return b.create<gpu::WaitOp>(loc, tokenType, deps).asyncToken();
   }
 
-  Value moveMemory(OpBuilder b, Value opr, uint32_t fidx, bool readAccess,
-                   bool writeAccess, llvm::SmallVector<Value> &copyBackOprs,
+  bool isOnDevice(ArrayRef<Operation *> oprUsers) const {
+    for (auto opUse : oprUsers) {
+      auto launch = dyn_cast<async::LaunchOp>(opUse);
+      // assumes the same GPU
+      if (!launch || !getGPUTarget(launch).has_value())
+        return false;
+    }
+    return true;
+  }
+
+  Value moveMemory(OpBuilder b, async::LaunchOp launchOp, Value opr,
+                   uint32_t fidx, bool readAccess, bool writeAccess,
+                   llvm::SmallVector<Value> &copyBackOprs,
                    llvm::SmallVector<Value, 8> &asyncDeps) const {
     Location loc = opr.getLoc();
     auto tokenType = b.getType<gpu::AsyncTokenType>();
     auto oprAllocOp = opr.getDefiningOp<memref::AllocOp>();
+    auto bAlloc = b;
     if (oprAllocOp)
-      b.setInsertionPoint(oprAllocOp);
+      bAlloc.setInsertionPoint(oprAllocOp);
 
-    Value allocWait = makeWait(b, loc);
+    Value allocWait = makeWait(bAlloc, loc);
     Type gpuMemType = opr.getType();
-    auto dst = b.create<gpu::AllocOp>(loc, gpuMemType, tokenType,
-                                      ValueRange{allocWait}, ValueRange{},
-                                      ValueRange{});
+    auto dst = bAlloc.create<gpu::AllocOp>(loc, gpuMemType, tokenType,
+                                           ValueRange{allocWait}, ValueRange{},
+                                           ValueRange{});
     Value dstMem = dst.getResult(0);
     Value dstToken = dst.getResult(1);
-    // if alloc, convert to gpu.alloc
-    if (oprAllocOp) {
-      // TODO(sjw): make sure accessors are all on the GPU
-      oprAllocOp->replaceAllUsesWith(ValueRange{dstMem});
-    } else {
-      if (readAccess) {
-        // else copy to device
-        auto memcpyToken = b.create<gpu::MemcpyOp>(
-            loc, tokenType, ValueRange{dstToken}, dstMem, opr);
-        dstToken = memcpyToken.getResult(0);
+
+    bool copyFlag = false;
+    auto makeCopy = [&]() {
+      if (!copyFlag) {
```

Comment:
I don't understand the copyFlag logic.  The idea is apparently to make sure that makeCopy() only makes a copy once, but I don't see any path in which makeCopy() is called more than once.

Reply:
Cleaned up. Yes, it was a left over from a previous process.